### ISSUE TITLE
refactor `@pre_process` to support `ui=docs` on method 

### ISSUE THREAD

**Issue by jorgejesus** [geopython/pygeoapi#414]
**Description**
The `@pre_process` decorator does a  header copy and format, checking before sending arguments to methods. On  this code point: 

```
https://github.com/geopython/pygeoapi/blob/master/pygeoapi/api.py#L197
```
It is necessary to pass `ui=docs` to the underline `def openapi(self, headers_, format_, openapi)`:

**Steps to Reproduce**
`def openapi()` is called when the openpai document is requested

**Expected behavior**
To pass `ui=docs` and generate the docs

**Screenshots/Tracebacks**

**Environment**
 - OS: linux
 - pygeoapi version: master 

**Additional context**
Dicussion from gitter:

```
@jorgejesus
Apr 18 19:31
Let me see if I understood, you want to intercept ui and add ui=docs ?!
And then pass it to the functions below the decorator ?
Tom Kralidis
@tomkralidis
Apr 18 19:59
Yup. For the OpenAPI endpoint
In pygeoapi.api
```

**Comment by jorgejesus** [comment]
@tomkralidis 

Pushed some code to my repository, the `@pre_process` seems to be working as expected (passing arguments).

- API REST end point with `ui` parameter [here](https://github.com/jorgejesus/pygeoapi/blob/6abd931286021473f511ee5ea5114f193368dd5a/pygeoapi/flask_app.py#L121) 
- The inside `pygeoapi.openapi` [here](https://github.com/jorgejesus/pygeoapi/blob/6abd931286021473f511ee5ea5114f193368dd5a/pygeoapi/api.py#L197)
- Only initial problems was the `json.dump` that had problems with the `openapi` object having datetime.
- Tested on `localhost:5000/openapi` and no issues aside from swarger expecting something else as dic object
- Tested on  other REST end points (collections et all) and no complains

**Comment by tomkralidis** [comment]
thanks @jorgejesus .  Comments:

- the web wrappers (like [`pygeoapi/flask_app.py`](https://github.com/geopython/pygeoapi/blob/master/pygeoapi/flask_app.py#L118-L119) and [`pygeoapi/starlette_app.py`](https://github.com/geopython/pygeoapi/blob/master/pygeoapi/starlette_app.py#L101-L102) should simply pass the dict of request parameters onto `pygeoapi.api`.  In other words, there should be no parameter handling at the web wrapper level, but in `pygeoapi.api`.  This allows for things like adding new web wrappers (Django, Pyramid, etc.) in an agnostic manner, as well as other frontend types that can simply pass request dicts to `pygeoapi.api` accordingly (like CLI tooling)
- given the request args are being passed from the web wrapper, yet I don't see them in https://github.com/geopython/pygeoapi/blob/master/pygeoapi/api.py#L197, so I am guessing request arguments are being intercepted by https://github.com/geopython/pygeoapi/blob/master/pygeoapi/api.py#L72 

Is there any way to update https://github.com/geopython/pygeoapi/blob/master/pygeoapi/api.py#L72 to also pass on request args?

**Comment by tomkralidis** [comment]
Ref, ensuing template once the above is implemented:
```diff
diff --git a/pygeoapi/templates/openapi.html b/pygeoapi/templates/openapi.html
index 74832db..17d0677 100644
--- a/pygeoapi/templates/openapi.html
+++ b/pygeoapi/templates/openapi.html
@@ -1,5 +1,29 @@
-<!-- HTML for static distribution bundle build -->
 <!DOCTYPE html>
+{% if data['ui'] == 'redoc' %}
+<html>
+  <head>
+    <title>ReDoc</title>
+    <!-- needed for adaptive design -->
+    <meta charset="utf-8"/>
+    <meta name="viewport" content="width=device-width, initial-scale=1">
+    <link href="https://fonts.googleapis.com/css?family=Montserrat:300,400,700|Roboto:300,400,700" rel="stylesheet">
+
+    <!--
+    ReDoc doesn't change outer page styles
+    -->
+    <style>
+      body {
+        margin: 0;
+        padding: 0;
+      }
+    </style>
+  </head>
+  <body>
+    <redoc spec-url='{{ data['openapi-document-path'] }}'></redoc>
+    <script src="https://cdn.jsdelivr.net/npm/redoc@next/bundles/redoc.standalone.js"> </script>
+  </body>
+</html>
+{% else %}
 <html lang="en">
   <head>
     <meta charset="UTF-8">
@@ -55,3 +79,4 @@
   <style>.swagger-ui .topbar .download-url-wrapper { display: none } undefined</style>
   </body>
 </html>
+{% endif %}
```

**Comment by jorgejesus** [comment]
to make it clear, the workflow and logic:

- E.g the `api_.openapi(request.headers, request.args,openapi)`  accepts the HTTP headers, the request.arg (something like `?redoc=HAMSTER`)  

https://github.com/geopython/pygeoapi/blob/master/pygeoapi/flask_app.py#L119

- The `@pre_format` will discart arg[1] (original headers) and arg[2] (request.arg), replacing them with `{'Content-Type': 'application/json', 'X-Powered-By': 'pygeoapi 0.7.0'}` and `html` (respectivamente), 

- All methods expect these new headers and formats (this is a problem is we change the args being passed from `@pre_format`)

- Currently to have `request.args`  on the "other side"  would be: `api_.openapi(request.headers, request.args,,request.args,openapi)` , since the 3rd arg is always passed (this is ugly), ,you can see example on  these 2 lines off my fork:
https://github.com/jorgejesus/pygeoapi/blob/c24b27ed8e490b68faa43bc24b6ec7493e395f87/pygeoapi/flask_app.py#L121
https://github.com/jorgejesus/pygeoapi/blob/c24b27ed8e490b68faa43bc24b6ec7493e395f87/pygeoapi/api.py#L199

https://github.com/jorgejesus/pygeoapi/blob/c24b27ed8e490b68faa43bc24b6ec7493e395f87/pygeoapi/api.py#L199

- Question, to we need to have the `X-Powered-By` headers on the `@pre_format` and not something after processing, and we would passe the original headers to the methods?
- I suggest the following change the code base, methods should more or less follow something like `func(cls, headers_, request_format_)`

Comments?

**Comment by jorgejesus** [comment]
After chatting with @tomkralidis:

* Perhaps pygeoapi.api.API ‘public’ methods all accept headers, request_args, path_params, type thing
* Code suggestion

```
import functools
def cool_decorator(func):
    @functools.wraps(func)
    def wrapper_decorator(*args, **kwargs):
        args2 = list(args)
        args2.pop(1)
        args2.pop(1)
        args2.insert(1, check_format(args[2], args[1]))  # add format
        args2.insert(1, HEADERS.copy())  # add headers
        args2.insert(3, args[2])
        value = func(*args2)
        return value
    return wrapper_decorator
```

**Comment by jorgejesus** [comment]
@tomkralidis 
I propose the following:

1. Implementation of an uniform  signature on `api.py`
Something like this ([link to code](https://github.com/jorgejesus/pygeoapi/blob/bd03a66f9c00eac08278950c33317cf6146b6393/pygeoapi/api.py#L106)): 
```def root(self, headers_ = None, args_ = None,format_ = None):```

Where headers, args and format are named arguments and set to default values, function's arguments would come first (since they are the most important and native to function) 
e.g: ```def openapi(self, openapi, headers_ = None, args_= None, format_= None):```

2. `@pre_process` decorator to look like: [this](https://github.com/jorgejesus/pygeoapi/blob/bd03a66f9c00eac08278950c33317cf6146b6393/pygeoapi/api.py#L73)

```

def pre_process(func):
    @functools.wraps(func) 
    def wrapper_pre_process(*args, **kwargs):
        kwargs["format_"] = check_format(kwargs["args_"],kwargs["headers_"])
        kwargs["headers_"] = HEADERS.copy()
        return func(*args,**kwargs)  
    return wrapper_pre_process
```
This simplifies the structure and we stop messing around with args.

3. flask_app.py would have [this](https://github.com/jorgejesus/pygeoapi/blob/bd03a66f9c00eac08278950c33317cf6146b6393/pygeoapi/flask_app.py#L141 ) signature structure:

```headers, status_code, content = api_.conformance(headers_ = request.headers, args_ = request.args)
```
4. Issues:

- Some functions dont require `args_` this could be checked by   `@pre_process` and hacked, but there are some problems on getting the function signature when chaining with `@jsonld decorator`

Basically what do you think of this `kwargs` approach ??? I find it a bit "cleaner"

**Comment by francbartoli** [comment]
@jorgejesus would it allow being extensible and issue for instance `ui=redoc`?

**Comment by tomkralidis** [comment]
Implemented in #720, leveraging the refactor and use of `pygeoapi.api.APIRequest`.
